crates/common/fork_choice/src/handlers.rs

```diff
@@ -235,3 +237,139 @@ pub fn on_attester_slashing(
+        current_epoch - 1
+    } else {
+        GENESIS_EPOCH
+    };
+    // If attestation target is from a future epoch, delay consideration until the epoch arrives
+    ensure!([current_epoch, previous_epoch].contains(&target.epoch));
+    Ok(())
+}
+
+pub fn validate_on_attestation(
+    store: &Store,
+    attestation: &Attestation,
+    is_from_block: bool,
+) -> anyhow::Result<()> {
+    let target = attestation.data.target;
+
+    // If the given attestation is not from a beacon block message, we have to check the target
+    // epoch scope.
+    if !is_from_block {
+        validate_target_epoch_against_current_time(store, attestation)?;
+    }
+
+    // Check that the epoch number and slot number are matching
+    ensure!(target.epoch == compute_epoch_at_slot(attestation.data.slot));
+
+    // Attestation target must be for a known block. If target block is unknown, delay consideration
+    // until block is found
+    ensure!(store.blocks.contains_key(&target.root));
+
+    // Attestations must be for a known block. If block is unknown, delay consideration until the
+    // block is found
+    ensure!(
+        store
+            .blocks
+            .contains_key(&attestation.data.beacon_block_root)
+    );
+    // Attestations must not be for blocks in the future. If not, the attestation should not be
+    // considered
+    ensure!(store.blocks[&attestation.data.beacon_block_root].slot <= attestation.data.slot);
+
+    // LMD vote must be consistent with FFG vote target
+    ensure!(
+        target.root
+            == get_checkpoint_block(store, attestation.data.beacon_block_root, target.epoch)
+    );
+
+    // Attestations can only affect the fork choice of subsequent slots.
+    // Delay consideration in the fork choice until their slot is in the past.
+    ensure!(store.get_current_slot() >= attestation.data.slot + 1);
+
+    Ok(())
+}
+
+pub fn store_target_checkpoint_state(store: &mut Store, target: Checkpoint) -> anyhow::Result<()> {
+    // Store target checkpoint state if not yet seen
+    if let std::collections::hash_map::Entry::Vacant(e) = store.checkpoint_states.entry(target) {
+        let mut base_state = store.block_states[&target.root].clone();
+        let target_slot = compute_start_slot_at_epoch(target.epoch);
+        if base_state.slot < target_slot {
+            base_state.process_slots(target_slot)?;
+        }
+        e.insert(base_state);
+    }
+
+    Ok(())
+}
+
+pub fn update_latest_messages(
+    store: &mut Store,
+    attesting_indices: Vec<u64>,
+    attestation: Attestation,
+) -> anyhow::Result<()> {
+    let target = attestation.data.target;
+    let beacon_block_root = attestation.data.beacon_block_root;
+    let mut non_equivocating_attesting_indices = vec![];
+
+    for &i in &attesting_indices {
+        if !store.equivocating_indices.contains(&i) {
+            non_equivocating_attesting_indices.push(i);
+        }
+    }
+
+    for i in &non_equivocating_attesting_indices {
```

Comment:
```suggestion
    for &index in &attesting_indices {
        if !store.equivocating_indices.contains(&index) {
            non_equivocating_attesting_indices.push(index);
        }
    }

    for index in &non_equivocating_attesting_indices {
```

I would use `index` instead of i